test/spec/exception-handling.wast

```diff
@@ -2,38 +2,110 @@
   (event $e1 (attr 0) (param i32 f32))
 
-  (func (export "throw_test")
-    (throw $e0 (i32.const 0))
+  (func (export "throw_single_value")
+    (throw $e0 (i32.const 5))
   )
 
-  (func (export "rethrow_test")
+  (func (export "throw_multiple_values")
+    (throw $e1 (i32.const 3) (f32.const 3.5))
+  )
+
+  (func (export "rethrow_nullref")
     (rethrow (ref.null))
   )
 
-  (func (export "try_test") (result i32)
+  (func (export "try_nothrow") (result i32)
     (try (result i32)
       (i32.const 3)
+      (catch
+        (drop (exnref.pop))
+        (i32.const 0)
+      )
+    )
+  )
+
+  (func (export "try_throw_catch") (result i32)
+    (try (result i32)
+      (throw $e0 (i32.const 5))
       (catch
         (drop (exnref.pop))
         (i32.const 3)
       )
     )
   )
 
-  (func (export "br_on_exn_test") (result i32)
+  (func (export "try_throw_rethrow")
+    (try
+      (throw $e0 (i32.const 5))
+      (catch
+        (rethrow (exnref.pop))
+      )
+    )
+  )
+
+  (func (export "br_on_exn_nullref") (result i32)
     (block $l0 (result i32)
       (drop
         (br_on_exn $l0 $e0 (ref.null))
       )
       (i32.const 0)
     )
   )
+
+  (func (export "br_on_exn_match_single_value") (result i32) (local $exn exnref)
+    (try (result i32)
+      (throw $e0 (i32.const 5))
+      (catch
+        (local.set $exn (exnref.pop))
+        (block $l0 (result i32)
+          (rethrow
+            (br_on_exn $l0 $e0 (local.get $exn))
+          )
+        )
+      )
+    )
+  )
+
+  (func (export "br_on_exn_match_multiple_values") (result i32 f32)
+        (local $exn exnref)
+    (try (result i32 f32)
+      (throw $e1 (i32.const 3) (f32.const 3.5))
+      (catch
+        (local.set $exn (exnref.pop))
+        (block $l0 (result i32 f32)
+          (rethrow
+            (br_on_exn $l0 $e1 (local.get $exn))
+          )
+        )
+      )
+    )
+  )
+
+  (func (export "br_on_exn_dont_match") (result i32) (local $exn exnref)
+    (try (result i32)
+      (throw $e1 (i32.const 3) (f32.const 3.5))
+      (catch
+        (local.set $exn (exnref.pop))
+        (block $l0 (result i32)
+          (rethrow
+            (br_on_exn $l0 $e0 (local.get $exn))
+          )
+        )
+      )
+    )
+  )
 )
 
-(assert_trap (invoke "throw_test"))
-(assert_trap (invoke "rethrow_test"))
-(assert_return (invoke "try_test") (i32.const 3))
-(assert_trap (invoke "br_on_exn_test"))
+(assert_trap (invoke "throw_single_value"))
+(assert_trap (invoke "throw_multiple_values"))
+(assert_trap (invoke "rethrow_nullref"))
+(assert_return (invoke "try_nothrow") (i32.const 3))
+(assert_return (invoke "try_throw_catch") (i32.const 3))
+(assert_trap (invoke "try_throw_rethrow"))
+(assert_trap (invoke "br_on_exn_nullref"))
+(assert_return (invoke "br_on_exn_match_single_value") (i32.const 5))
+(assert_return (invoke "br_on_exn_match_multiple_values") (tuple.make (i32.const 3) (f32.const 3.5)))
+(assert_trap (invoke "br_on_exn_dont_match"))
```

Comment:
Another test to add would be throwing and catching across a function boundary, if that works.

Reply:
Sounds good. Add several more tests to test this (throw in A and catch in B, throw in A in rethrow in B and catch in C, ...).